src/app/server/Dnssd.cpp

```diff
@@ -412,13 +370,27 @@ void DnssdServer::StartServer(Dnssd::CommissioningMode mode)
 #if CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY
     else if (GetExtendedDiscoveryTimeoutSecs() != CHIP_DEVICE_CONFIG_DISCOVERY_DISABLED)
     {
-        err = AdvertiseCommissionableNode(mode);
-        if (err != CHIP_NO_ERROR)
+        bool alwaysAdvertiseExtended = (GetExtendedDiscoveryTimeoutSecs() == CHIP_DEVICE_CONFIG_DISCOVERY_NO_TIMEOUT);
+        // We do extended discovery advertising in three cases:
+        // 1) We don't have a timeout for extended discovery.
+        // 2) We are transitioning out of commissioning mode (basic or enhanced)
```

Comment:
Are you sure we shouldn't start the extended discovery when the device starts with the commissioning mode off? I thought Pairing Hint could include instructions on how to put the device in the commissioning mode so from that perspective it would be useful to have a way to start the advertising with CM=0.

Reply:
@Damian-Nordic I asked @chrisdecenzo that question and this implements what I believe he explained as the goal.

That said, I can see the argument for doing extended discovery if we have no fabrics commissioned and are just starting up.  I can modify this PR to make that work.  Or do you think we should also advertise extended discovery on startup even if there are fabrics commissioned?

Reply:
Yeah, based on the description of **CommissioningModeInitialStepsHint** I do think we should advertise extended discovery on startup:
> This field SHALL identify a hint for the steps that MAY be used to put a device that has not yet been
commissioned into commissioning mode.
> (...)
> Devices that implement Extended Discovery SHALL reflect this value in the Pairing Hint field of Commissionable Node Discovery when they have not yet been commissioned.

Would you agree?

Reply:
@Damian-Nordic raises a good point I had not considered

Reply:
Yep, done.